### PR TITLE
Resultset distinguishable

### LINKED ISSUE #468: resultset are indistinguishable from pandas data frames
ResultSet (the data structure we use by default to store db results), is currently indistinguishable from pandas data frames:

image

this can confuse users, as they might think the're data frames. we should ensure they look different. perhaps add result set at the bottom.

maybe we should also detect when they try do perform pandas-like operations? e.g.,

```
df.head()
```

Error should be: `head is not a valid operation, you can convert this into a pandas data frame by calling .DataFrame() or a polars data frame by calling .PolarsDataFrame()`

### PR DESCRIPTION
**Pull request by bbeat2782**
## Describe your changes

Adding description after html representation of `ResultSet` and error message for AttributeError when invalid operation is performed

## Issue number

Closes #468 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--655.org.readthedocs.build/en/655/

### CODE REVIEW

**Review by edublancas — CHANGES_REQUESTED** [review]
overall, it looks good, can we make the font smaller? it's taking too much space

image

**Comment by bbeat2782** [comment]
> overall, it looks good, can we make the font smaller? it's taking too much space
> 
> image

Sure. Should I keep the `ResultSet` at the front as it is and reduce the size of text that comes after `ResultSet` or should I reduce all of them?

**Comment by edublancas** [comment]
please reduce all of them from ResultSet until PolarsDataFrame()

File: src/tests/test_resultset.py [hunk 1]
```diff
@@ -78,6 +78,20 @@ def test_resultset_repr_html(result_set):
         "<tr>\n            <td>0</td>\n        </tr>\n        <tr>\n            "
         "<td>1</td>\n        </tr>\n        <tr>\n            <td>2</td>\n        "
         "</tr>\n    </tbody>\n</table>"
+        "<div><code>ResultSet</code> : "
+        "to convert to pandas, call <code>.DataFrame()</code> "
+        "or to polars, call <code>.PolarsDataFrame()</code></div>"
+    )
+
+
+def test_invalid_operation_error(result_set):
```
Comment (yafimvo) [inline src/tests/test_resultset.py:87]: Can we use parameters here and test different invalid functions (e.g. `tail`, `head(-1)`, `value_counts`, `not_df_function`)?

File: src/sql/run.py [hunk 2]
```diff
@@ -179,6 +184,18 @@ def __getitem__(self, key):
...
+        "Raises AttributeError when invalid operation is performed."
+        try:
+            return object.__getattribute__(self, attr)
+        except AttributeError:
+            err_msg = (
+                f"'{attr}' is not a valid operation, you can convert this "
+                "into a pandas data frame by calling '.DataFrame()' or a "
+                "polars data frame by calling '.PolarsDataFrame()'"
+            )
+            raise AttributeError(err_msg) from None
```
Comment (yafimvo) [inline src/sql/run.py:197]: @edublancas 
Shouldn't we use `UsageError` here?

**Review by yafimvo — CHANGES_REQUESTED** [review]
lgtm. added some minor notes.

General thought:
image

I agree with @edublancas, I'm not sure the bottom message is clear enough. Maybe we can style it as the `Truncated` message (smaller font and italic style)?

File: src/sql/run.py [hunk 3]
```diff
@@ -126,6 +126,11 @@ def _repr_html_(self):
         if self.pretty:
             self.pretty.add_rows(self)
             result = self.pretty.get_html_string()
+            result += (
```
Comment (yafimvo) [inline src/sql/run.py:129]: can we use the string formatting instead of +?

**Review by yafimvo — CHANGES_REQUESTED** [review]
@AnirudhVIyer Also, please rebase.

File: src/sql/run.py (hunk 2 above)
Comment (edublancas) [reply src/sql/run.py:200]: AttributeError makes sense here to keep compatibility with the existing Python error

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by bbeat2782** [comment]
Changes are
1. Added text under `ResultSet` representation is now smaller and italicized
2. Test function now tests different pandas-like functions with `@pytest.mark.parametrize`
3. Use string formatting instead of + when adding 1)

If we want smaller text for 1), I will make the change.

File: src/sql/run.py [hunk 4]
```diff
@@ -126,6 +126,14 @@ def _repr_html_(self):
         if self.pretty:
             self.pretty.add_rows(self)
             result = self.pretty.get_html_string()
+            EXTRA = (
+                "%s\n<span style='font-style:italic;font-size:small'>"
+                "<code>ResultSet</code> : "
+                "to convert to pandas, call <code>.DataFrame()</code> "
+                "or to polars, call <code>.PolarsDataFrame()</code></span><br>"
```
Comment (edublancas) [inline src/sql/run.py:133]: looks better, let's make it a bit smaller

also, let's add some links. click on DataFrame() should take to https://jupysql.ploomber.io/en/latest/integrations/pandas.html

and clicking on PolarsDataFrame should take to https://jupysql.ploomber.io/en/latest/integrations/pandas.html

**Review by edublancas — CHANGES_REQUESTED** [review]

**Review by yafimvo — APPROVED** [review]

**Review by yafimvo — COMMENTED** [review]
I think it looks good.

Why do we get this message here? is this the expected behavior?

image

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
@yafimvo: good points. I'll open issues to follow up in your observations

**Pull request merged**